pteranodon/autoconnect.py

```diff
@@ -0,0 +1,234 @@
+        :type server_ip_addr: Optional[str]
+        """
+        self._server_addr = server_ip_addr
+        self._should_check_server = self._server_addr is not None
+
+    async def _test_port_open(
+        self,
+        port: int,
+        socket_kind: int,
+        timeout: float = 1.0,
+    ) -> Optional[Tuple[int, bool]]:
+        """
+        Attempts to connect to a port with a specified timeout and socket kind to determine if the port is open.
+
+        :param port: The port to check
+        :type port: int
+        :param socket_kind: The type of socket to open (SOCK_STREAM for TCP, SOCK_DGRAM for UDP)
+        :type socket_kind: socket.SocketKind
+        :param timeout: The timeout for the connect call
+        :type timeout: float
+        :return: The port tested and whether the port is open, None if invalid parameters
+        :rtype: Optional[Tuple[int, bool]]
+        """
+        # Must have a server address
+        if not self._should_check_server:
+            return None
+
+        # Only supporting TCP and UDP
+        if socket_kind not in [socket.SOCK_STREAM, socket.SOCK_DGRAM]:
+            return None
+
+        with socket.socket(socket.AF_INET, socket_kind) as sock:
+            sock.settimeout(timeout)
+            try:
+                sock.connect((self._server_addr, port))
+            except TimeoutError:
+                return port, False
+
+        return port, True
+
+    def fetch_possible_serial_devs(self) -> List[str]:
+        """
+        Fetches all the possible serial devices from /sys/class/tty while removing standard devices and virtual consoles.
+
+        :return: The paths of connected serial devices
+        :rtype: List[str]
+        """
+        blacklisted_devices = ["pmtx", "console", "ttyprintk"]
+
+        serial_devices = os.listdir("/sys/class/tty")
+
+        # Remove blacklisted files
+        serial_devices = [
+            dev for dev in serial_devices if dev not in blacklisted_devices
+        ]
+
+        # Remove virtual consoles (matches "tty" followed by any number and nothing else)
+        serial_devices = [
+            dev for dev in serial_devices if re.match("(tty)[0-9]*$", dev) is None
+        ]
+
+        serial_paths = ["/dev/" + dev for dev in serial_devices]
+
+        return serial_paths
+
+    # Must be done in parallel, would take 17+ hours otherwise
+    def fetch_open_proto_ports(self, socket_kind: int) -> List[int]:
+        """
+        Fetches the open ports for the given protocol (not necessarily open for MAVSDK). Blocking.
+
+        :param socket_kind: The kind of socket to test (SOCK_STREAM for TCP, SOCK_DGRAM for UDP)
+        :type socket_kind: socket.SocketKind
+        :return: A list of open UDP or TCP ports
+        :rtype: List[int]
+        """
+        # Only supporting TCP and UDP
+        if socket_kind not in [socket.SOCK_STREAM, socket.SOCK_DGRAM]:
+            return []
+
+        port_range = range(1024, pow(2, 16))
```

Comment:
Only need to check valid MAVSDK ports, which would actually start at 14540 and realistically go to some range specified in the function call. So 14540-14550 for example. A given computer most likely does not host lots of drones, and this fails if it is since multiple connections to a server override the first.

Reply:
So I understand your comment on the lower port bound, although where is that defined in the MAV SDK docs? I know that that is the default port but can't a MAV SDK server theoretically run on any open port so I could start one at 12000, which would be perfectly reasonable? 

I am also confused by the last comment. One important part is that this function checks just that the port is open to TCP/UDP, not that MAV SDK is running, which still needs to be checked. This also returns a list of ports so I am confused about how multiple connections to a server override the first.